packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart

```diff
@@ -188,21 +215,113 @@ class FuchsiaDevice extends Device {
 
   @override
   Future<LaunchResult> startApp(
-    ApplicationPackage package, {
+    covariant FuchsiaApp package, {
     String mainPath,
     String route,
     DebuggingOptions debuggingOptions,
     Map<String, dynamic> platformArgs,
     bool prebuiltApplication = false,
     bool usesTerminalUi = true,
     bool ipv6 = false,
-  }) =>
-      Future<void>.error('unimplemented');
+  }) async {
+    if (!prebuiltApplication) {
+      await buildFuchsia(fuchsiaProject: FlutterProject.current().fuchsia,
+                         target: mainPath,
+                         buildInfo: debuggingOptions.buildInfo);
+    }
+    // Stop the app if it's currently running.
+    await stopApp(package);
+    // Find out who the device thinks we are.
+    final String host = await fuchsiaSdk.fuchsiaDevFinder.resolve(name);
+    final int port = await os.findFreePort();
+    if (port == 0) {
+      printError('Failed to find a free port');
+      return LaunchResult.failed();
+    }
+    final Directory packageRepo =
+        fs.directory(fs.path.join(getFuchsiaBuildDirectory(), '.pkg-repo'));
+    packageRepo.createSync(recursive: true);
+
+    final String appName = FlutterProject.current().manifest.appName;
+
+    FuchsiaPackageServer fuchsiaPackageServer;
+    bool serverRegistered = false;
+    try {
```

Comment:
I would use a status spinner here if this takes longer than a second or so

```
Status buildStatus = logger.startProgress('Serving fuchsia mod...', timeout: null);
...
finally {
  buildStatus.stop/cancel/dont remember
}


```

Reply:
Good idea. Done.